Source/MOLAuthenticatingURLSession.m

```diff
@@ -242,69 +271,71 @@ - (NSURLCredential *)clientCredentialForProtectionSpace:(NSURLProtectionSpace *)
     [self log:@"Server Trust: No server trust information available"];
     return nil;
   }
 
   OSStatus err = errSecSuccess;
 
-  if (self.serverRootsPemFile) {
-    NSError *error;
-    NSData *rootsData = [NSData dataWithContentsOfFile:self.serverRootsPemFile
-                                               options:0
-                                                 error:&error];
-    if (!rootsData) {
-      [self log:@"Unable to read server root certificate file %@ with error: %@",
-          self.serverRootsPemFile, error.localizedDescription];
-      return nil;
+  // A local SecTrustRef to store custom anchors if they exist. SecTrustSetAnchorCertificates in
+  // combination with SecTrustEvaluate will leak a SecTrustRef if we were to set the anchors on
+  // protectionSpace.serverTrust. If there are custom anchors serverTrust will be evaluated, if
+  // there are no custom anchors protectionSpace.serverTrust will be evaluated.
+  SecTrustRef serverTrust = NULL;
+
+  if (self.anchors) {
+    CFArrayRef policies = NULL;
+    SecTrustCopyPolicies(protectionSpace.serverTrust, &policies);
+    int certCount = (int)SecTrustGetCertificateCount(protectionSpace.serverTrust);
+    NSMutableArray *serverTrustCertRefs = [[NSMutableArray alloc] initWithCapacity:certCount];
+    for (int i = 0; i < certCount; i++) {
+      [serverTrustCertRefs addObject:
+          (__bridge id)SecTrustGetCertificateAtIndex(protectionSpace.serverTrust, i)];
     }
-    self.serverRootsPemData = rootsData;
-  }
-
-  if (self.serverRootsPemData) {
-    NSString *pemStrings = [[NSString alloc] initWithData:self.serverRootsPemData
-                                                 encoding:NSASCIIStringEncoding];
-    NSArray *certs = [MOLCertificate certificatesFromPEM:pemStrings];
 
-    // Make a new array of the SecCertificateRef's from the MOLCertificate's.
-    NSMutableArray *certRefs = [[NSMutableArray alloc] initWithCapacity:certs.count];
-    for (MOLCertificate *cert in certs) {
-      [certRefs addObject:(id)cert.certRef];
-    }
+    // Create a copy of protectionSpace.serverTrust by grabbing it's policies and certificates.
+    SecTrustCreateWithCertificates((__bridge CFTypeRef)serverTrustCertRefs, policies, &serverTrust);
+    if (policies) CFRelease(policies);
 
     // Set this array of certs as the anchors to trust.
-    err = SecTrustSetAnchorCertificates(serverTrust, (__bridge CFArrayRef)certRefs);
+    err = SecTrustSetAnchorCertificates(serverTrust, (__bridge CFArrayRef)self.anchors);
     if (err != errSecSuccess) {
       [self log:@"Server Trust: Could not set anchor certificates: %d", err];
+      if (serverTrust) CFRelease(serverTrust);
       return nil;
     }
   }
 
   // Print details about the server's leaf certificate.
-  SecCertificateRef firstCert = SecTrustGetCertificateAtIndex(serverTrust, 0);
+  SecCertificateRef firstCert =
+      SecTrustGetCertificateAtIndex(serverTrust ?: protectionSpace.serverTrust, 0);
```

Comment:
You can avoid the elvis check here; given that the certs are copied from protectionSpace.serverTrust to serverTrust you can just use protectionSpace.serverTrust here.